hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java

```diff
@@ -2904,7 +2908,8 @@ public boolean processReport(final DatanodeID nodeID,
       }
       if (namesystem.isInStartupSafeMode()
           && !StorageType.PROVIDED.equals(storageInfo.getStorageType())
-          && storageInfo.getBlockReportCount() > 0) {
+          && storageInfo.getBlockReportCount() > 0
+          && totalReportNum == currentReportNum) {
```

Comment:
If a datanode report twice during namenode safemode, the second report will be almost completely processed, which may extend startup time. How about modify code like this? This can also avoid changes in the method signature.
```
if (namesystem.isInStartupSafeMode()
          && !StorageType.PROVIDED.equals(storageInfo.getStorageType())
          && storageInfo.getBlockReportCount() > 0) {
        blockLog.info("BLOCK* processReport 0x{} with lease ID 0x{}: "
            + "discarded non-initial block report from datanode {} storage {} "
            + " because namenode still in startup phase",
            strBlockReportId, fullBrLeaseId, nodeID, storageInfo.getStorageID());
        boolean needRemoveLease = true;
        for (DatanodeStorageInfo sInfo : node.getStorageInfos()) {
          if (sInfo.getBlockReportCount() == 0) {
            needRemoveLease = false;
          }
        }
        if (needRemoveLease) {
          blockReportLeaseManager.removeLease(node);
        }
        return !node.hasStaleStorages();
      }
```

Reply:
@zhangshuyan0 Thank you for your retrial,This change can achieve the same effect, but I think node.hasStaleStorages() is also a Datanode-level operation that should also be called on the last disk, but logically, functionally, it's not that different。Listen to other people's opinions ，@Hexiaoqiao What do you think about that

Reply:
I like @zhangshuyan0's proposal better. 

The following section of code can also be separated to be a function on its own.
```
// Remove the lease when we have received block reports for all storages for a particular DN.
void removeLease() {

        for (DatanodeStorageInfo sInfo : node.getStorageInfos()) {
          if (sInfo.getBlockReportCount() == 0) {
            needRemoveLease = false;
          }
        }
        if (needRemoveLease) {
          blockReportLeaseManager.removeLease(node);
        }
}
```

Reply:
I am not sure if this will be a good solution with condition `blockReportCount == 0`, consider that one disk failed but not checked in time. Will it affect this logic here? Thanks.

Reply:
Two of the cases will come into this logic
1. The namenode is restarting while receiving FBRS from all datanodes and is in safe mode
2. When the namenode is in secure mode for some reason while it has been running for a long time
In the first case, if the datanode has a failed disk, the datanode will send the FBR for the normal disk and the namenode will handle it normally
In the second case, blockReportCount == 0 will always be false if no new disks are added to the datanode
So I recommend keeping the code as it is and not using blockReportCount == 0